pkg/storaged/stratis/pool.jsx

```diff
@@ -64,6 +65,291 @@ const _ = cockpit.gettext;
+ *
+ * { slot: number | null, keydesc: string }
+ *
+ * A V1 pool will have at most one entry in the array, and it will
+ * have `slot == null`.  V2 pools have zero or more entries with
+ * numbered slots.
+ */
+
+function get_key_descriptions(pool) {
+    const result = [];
+
+    if (!pool.Encrypted)
+        return result;
+
+    if (client.stratis_interface_revision < 8) {
+        const key_desc = pool.KeyDescription;
+        /* The KeyDescription property has type "(b(bs))".
+
+           The first "b" is false when the encryption of the pool is
+           inconsistent (not all block devices have the same LUKS
+           header). We pretend that the pool doesn't have any
+           passphrase in that case.
+
+           The second "b" is true when the pool has a passphrase, and
+           the "s" is the description of the key for it.
+
+           Thus, we detect an object of shape
+
+               [ true, [ true, "..." ] ]
+
+           and fish the string out of it.
+         */
+        if (key_desc[0] && key_desc[1][0])
+            result.push({ slot: null, keydesc: key_desc[1][1] });
+    } else {
+        let key_descs = pool.KeyDescriptions;
+        /* The KeyDescriptions property itself has type "v". For a V1
+           pool this variant carries a value of type "b(bs)" just as
+           with the r6 API (see above). For a V2 pool, it carries a
+           "a(us)" value, an array of "slot number plus key
+           description".
+
+           Since the property itself has type "v", we need unwrap this
+           variant explicitly.
+         */
+
+        /* HACK - https://github.com/stratis-storage/stratisd/issues/3895
+
+           Introduced in stratisd 3.8.0, fixed in stratisd 3.8.3.
+
+           Change notifications drop the variant wrapping for some
+           reason, so we only do the unwrapping when "key_descs" is
+           indeed a variant.
+         */
+        if ("t" in key_descs)
+            key_descs = key_descs.v;
+        if (pool.MetadataVersion == 1) {
+            if (key_descs[0] && key_descs[1][0])
+                result.push({ slot: null, keydesc: key_descs[1][1] });
+        } else if (pool.MetadataVersion == 2) {
+            for (const key_desc of key_descs)
+                result.push({ slot: key_desc[0], keydesc: key_desc[1] });
+        }
+    }
+
+    return result;
+}
+
+/* Clevis information is returned as an array with elements of type
+ *
+ *   { slot: number | null, pin: "tang", url: string }
+ * | { slot: number | null, pin: string }
+ *
+ * A V1 pool will have at most one entry in the array, and it will
+ * have `slot == null`.  V2 pools have zero or more entries with
+ * numbered slots.
+ */
+
+function get_clevis_infos(pool) {
+    const result = [];
+
+    if (!pool.Encrypted)
+        return result;
+
+    if (client.stratis_interface_revision < 8) {
+        const clevis_info = pool.ClevisInfo;
+        /* The ClevisInfo property has type "(b(b(ss)))".
```

Comment:
Really weird types. I believe you, but what's the point of putting a tuple inside of a tuple inside of a tuple.. may as well just be flat.

Reply:
*shrug* It's probably from mechanical encoding of Rust `Option` types.